packages/gatsby-image/src/index.js

```diff
@@ -362,6 +362,31 @@ class Image extends React.Component {
     this.placeholderRef = props.placeholderRef || React.createRef()
     this.handleImageLoaded = this.handleImageLoaded.bind(this)
     this.handleRef = this.handleRef.bind(this)
+
+    // Supports Art Direction feature to correctly render image variants
+    const imageVariants = props.fluid || props.fixed
+    if (hasArtDirectionSupport(imageVariants)) {
+      // When image variant changes, adapt via render immediately instead of delaying
+      // until new image request triggers 'onload' event.
+      // Can be verified on 'slow 3G' with no initial cache.
+      if (isBrowser) {
+        this.handleVariantChange = this.handleVariantChange.bind(this)
+        const base = imageVariants.find(v => !v.media)
+
+        this.mq = { base }
+        this.mq.queries = imageVariants.filter(v => v.media)
+        this.mq.queries.forEach(v => {
+          this.mq[v.media] = v
+          const mql = window.matchMedia(v.media)
+          if (mql.addEventListener) {
+            mql.addEventListener(`change`, this.handleVariantChange)
+          } else {
+            // Deprecated 'MediaQueryList' API, <Safari 14, IE, <Edge 16
+            mql.addListener(this.handleVariantChange)
```

Comment:
We probably want to clean it up with `componentWillUnmount`. Besides that this looks great! Thank you!

Reply:
I'm happy to take over as you've been waiting long on this feedback.